cli/src/commands/workspace.rs

```diff
@@ -279,6 +282,76 @@ fn cmd_workspace_root(
     Ok(())
 }
 
+fn create_and_check_out_recovery_commit(
+    command: &CommandHelper,
+    ui: &mut Ui,
+    workspace: Workspace,
+) -> Result<Arc<ReadonlyRepo>, CommandError> {
+    let operation = command.resolve_operation(ui, workspace.repo_loader())?;
+    let repo = workspace.repo_loader().load_at(&operation)?;
+    let workspace_id = workspace.workspace_id().clone();
+    let tree_id = repo.store().empty_merged_tree_id();
+
+    let mut workspace_command = command.for_loaded_repo(ui, workspace, repo.clone())?;
```

Comment:
This is basically the same as `command.workspace_helper_no_snapshot()`?
Can be something like
```rust
    let mut workspace_command = command.workspace_helper_no_snapshot(ui)?;
    let workspace_id = workspace_command.workspace_id().clone();
    let mut tx = workspace_command.start_transaction().into_inner();
```

Reply:
Ah, good catch. I used your suggestion, except that for the last line, I had to use `repo.start_transaction(command.settings())` instead to avoid borrow checker problems (before this line, I call `workspace_command.unchecked_start_working_copy_mutation()`, which also borrows `workspace_command` mutably).

Reply:
It's a bit tricky, but these lines can be reordered. `.start_transaction().into_inner()` releases `&mut workspace_command`.

`repo.start_transaction()` doesn't record the command arguments in the operation. You can see the difference in `jj op log`.

Reply:
Ah, I did not realize this. Let me send a PR to correct this.